src/services/ledger/index.ts

```diff
@@ -305,31 +317,78 @@ export const LedgerService = (): ILedgerService => {
   })
 }
 
-export const translateToLedgerTx = (tx): LedgerTransaction => ({
-  id: tx.id,
-  walletId: toWalletId(tx.accounts),
-  type: tx.type,
-  debit: toSats(tx.debit),
-  credit: toSats(tx.credit),
-  fee: toSats(tx.fee),
-  usd: tx.usd,
-  feeUsd: tx.feeUsd,
-  currency: tx.currency,
-  timestamp: tx.timestamp,
-  pendingConfirmation: tx.pending,
-  journalId: tx._journal.toString(),
-  lnMemo: tx.memo,
-  username: tx.username,
-  memoFromPayer: tx.memoPayer,
-  paymentHash: tx.hash,
-  pubkey: tx.pubkey,
-  address:
-    tx.payee_addresses && tx.payee_addresses.length > 0
-      ? tx.payee_addresses[0]
-      : undefined,
-  txHash: tx.hash,
-  feeKnownInAdvance: tx.feeKnownInAdvance || false,
-})
+const translateToLedgerTx = ({
+  txnRecord,
+  txnMetadata,
+}: {
+  txnRecord: TransactionRecord
+  txnMetadata: LedgerTransactionMetadata | undefined
+}): LedgerTransaction => {
+  const fromTxnRecord = {
+    id: txnRecord.id as LedgerTransactionId,
+    walletId: toWalletId(txnRecord.accounts as LiabilitiesWalletId),
+    type: txnRecord.type as LedgerTransactionType,
+    debit: toSats(txnRecord.debit),
+    credit: toSats(txnRecord.credit),
+    fee: toSats(txnRecord.fee),
+    usd: txnRecord.usd,
+    feeUsd: txnRecord.feeUsd,
+    currency: txnRecord.currency as WalletCurrency,
+    timestamp: txnRecord.timestamp,
+    pendingConfirmation: txnRecord.pending,
+    journalId: txnRecord._journal.toString(),
+    lnMemo: txnRecord.memo,
+    username: (txnRecord.username as Username) || undefined,
+    memoFromPayer: txnRecord.memoPayer,
+    paymentHash: (txnRecord.hash as PaymentHash) || undefined,
+    pubkey: (txnRecord.pubkey as Pubkey) || undefined,
+    address:
+      txnRecord.payee_addresses && txnRecord.payee_addresses.length > 0
+        ? (txnRecord.payee_addresses[0] as OnChainAddress)
+        : undefined,
+    txHash: (txnRecord.hash as OnChainTxHash) || undefined,
+    feeKnownInAdvance: txnRecord.feeKnownInAdvance || false,
+  }
+
+  if (txnMetadata === undefined) return fromTxnRecord
+
+  let fromTxnMetadata: PartialLedgerTransactionFromMetadata = {}
+  if ("revealedPreImage" in txnMetadata) {
+    fromTxnMetadata = { revealedPreImage: txnMetadata.revealedPreImage }
+  }
+  return { ...fromTxnRecord, ...fromTxnMetadata }
+}
+
+const translateTxRecordsToLedgerTxs = async (
+  txnRecords: TransactionRecord[],
+): Promise<LedgerTransaction[]> => {
+  const txnsMetadata = await Promise.all(
+    txnRecords.map(async (tx): Promise<LedgerTransactionMetadata | undefined> => {
+      const txnMetadataResult = await txnMetadataRepo.findById(
```

Comment:
We could, it would be the same as implementing a similar join [to this](https://github.com/GaloyMoney/galoy/pull/1044/files#diff-463de0ec8400f9be4647cb42dee40f257b5055c84cbe9328daa260760f4a1904R285) here in-code (it look slike a `View` is just an aggregation pipeline [computed on-demand](https://www.mongodb.com/docs/manual/core/views/)).

Reply:
We could, it would be the same as implementing a similar join [to this](https://github.com/GaloyMoney/galoy/pull/1044/files#diff-463de0ec8400f9be4647cb42dee40f257b5055c84cbe9328daa260760f4a1904R285) here in-code (it looks like a `View` is just an aggregation pipeline [computed on-demand](https://www.mongodb.com/docs/manual/core/views/)).

Reply:
Ok I implemented here as an option: https://github.com/GaloyMoney/galoy/compare/pass-preimage-to-ledger-transactions...optimize-metadata-join

It's not ideal though because we're moving away from using the `Mainbook.ledger` query which means we'd have to ensure it's faithfully reproduced.

I'd be in favour of merging this version of the code and then profiling how much longer the extra `transactions` gql query takes in staging before we consider optimizing it (or reverting if there's no good solution) like we suggested in #1044